File: src/worldbook.js
/**
 * 世界书管理模块
 * 兼容 SillyTavern WorldInfo 格式
 */

import fs from 'fs';
import path from 'path';
import { safeJsonParse, safeJsonParseWithFallback } from './json-utils.js';

export class WorldBookManager {
    constructor(dataDir) {
        this.dataDir = dataDir;
        this.worldsDir = path.join(dataDir, 'worlds');
        this.cache = new Map();
        this.currentWorldBook = null;
        this.currentWorldBookName = null;
        
        // ✅ 确保世界书目录存在
        if (!fs.existsSync(this.worldsDir)) {
            fs.mkdirSync(this.worldsDir, { recursive: true });
        }
    }


    /**
     * 读取世界书
     */
    readWorldBook(characterName) {
        // 检查缓存
        if (this.cache.has(characterName)) {
            return this.cache.get(characterName);
        }

        // 尝试多种可能的文件名格式
        const possibleNames = [
            characterName + "'s Lorebook.json",  // 弯引号
            characterName + "'s Lorebook.json",  // 直引号
            characterName + " Lorebook.json",
            characterName + ".json"
        ];

        // 先尝试直接匹配
        for (const name of possibleNames) {
            const filePath = path.join(this.worldsDir, name);
            if (fs.existsSync(filePath)) {
                const worldBook = safeJsonParse(fs.readFileSync(filePath, 'utf8'), 10 * 1024 * 1024); // 世界书限制 10MB
                this.cache.set(characterName, worldBook);
                return worldBook;
            }
        }

        // 模糊匹配：扫描目录查找包含角色名的文件
        if (fs.existsSync(this.worldsDir)) {
            const files = fs.readdirSync(this.worldsDir);
            for (const file of files) {
                if (file.includes(characterName) && file.endsWith('.json')) {
                    const filePath = path.join(this.worldsDir, file);
                    const worldBook = safeJsonParse(fs.readFileSync(filePath, 'utf8'), 10 * 1024 * 1024);
                    this.cache.set(characterName, worldBook);
                    return worldBook;
                }
            }
        }

        return null;
    }

    /**
     * 匹配世界书条目
     * @param {Object} worldBook - 世界书对象
     * @param {string} inputText - 用于匹配的文本（包括历史消息）
     * @param {number} maxEntries - 最大返回条目数
     * @param {Set<string>} stickyKeys - 当前会话中仍然粘性的条目键集合
     */
    matchEntries(worldBook, inputText, maxEntries = 10, stickyKeys = new Set()) {
        if (!worldBook || !worldBook.entries) {
            return [];
        }

        const normalizeKeys = (value) => {
            if (Array.isArray(value)) {
                return value.filter((item) => typeof item === 'string' && item.trim());
            }

            if (typeof value === 'string' && value.trim()) {
                return [value];
            }

            return [];
        };

        const isEnabled = (value, fallback = true) => {
            if (value === undefined || value === null || value === '') {
                return fallback;
            }

            if (typeof value === 'string') {
                const normalized = value.trim().toLowerCase();
                if (['false', '0', 'off', 'no'].includes(normalized)) {
                    return false;
                }
                if (['true', '1', 'on', 'yes'].includes(normalized)) {
                    return true;
                }
            }

            return Boolean(value);
        };

        const normalizeNumber = (value, fallback = 0) => {
            const normalized = Number(value);
            return Number.isFinite(normalized) ? normalized : fallback;
        };

        const normalizePosition = (value) => {
            if (typeof value === 'string') {
                const normalized = value.trim().toLowerCase();
                if ([
                    '1',
                    'after',
                    'after_char',
                    'after_character',
                    'after_description',
                    'post',
                    'post_history',
                    'post-history'
                ].includes(normalized)) {
                    return 1;
                }

                if ([
                    '0',
                    'before',
                    'before_char',
                    'before_character',
                    'before_description',
                    'system',
                    'pre_history',
                    'pre-history'
                ].includes(normalized)) {
                    return 0;
                }
            }

            return normalizeNumber(value, 0);
        };

        const matched = [];
        const constants = [];
        const stickyMatched = [];
        const inputLower = inputText.toLowerCase();

        const entriesArray = Array.isArray(worldBook.entries)
            ? worldBook.entries
            : Object.values(worldBook.entries);

        for (const entry of entriesArray) {
            if (!isEnabled(entry.enabled, true) || isEnabled(entry.disable, false)) {
                continue;
            }

            const keys = normalizeKeys(entry.keys ?? entry.key);
            const entryKey = entry.uid || entry.id || keys[0] || entry.comment || entry.name || 'unknown';
            const sticky = normalizeNumber(entry.sticky, 0);
            const order = normalizeNumber(entry.order ?? entry.insertion_order, 0);
            const position = normalizePosition(entry.position);

            if (isEnabled(entry.constant, false)) {
                constants.push({
                    content: entry.content,
                    order,
                    key: entryKey,
                    isConstant: true,
                    position,
                    sticky: 0
                });
                continue;
            }

            const isStickyActive = stickyKeys.has(entryKey);
            const secondaryKeys = normalizeKeys(entry.secondary_keys ?? entry.keysecondary);
            let primaryMatch = false;
            let secondaryMatch = secondaryKeys.length === 0;

            for (const key of keys) {
                if (key && inputLower.includes(key.toLowerCase())) {
                    primaryMatch = true;
                    break;
                }
            }

            if (primaryMatch && secondaryKeys.length > 0) {
                const logic = normalizeNumber(entry.selectiveLogic, 0);

                if (logic === 0) {
                    secondaryMatch = secondaryKeys.some((k) => k && inputLower.includes(k.toLowerCase()));
                } else if (logic === 1) {
                    secondaryMatch = !secondaryKeys.every((k) => k && inputLower.includes(k.toLowerCase()));
                } else if (logic === 2) {
                    secondaryMatch = !secondaryKeys.some((k) => k && inputLower.includes(k.toLowerCase()));
                } else if (logic === 3) {
                    secondaryMatch = secondaryKeys.every((k) => k && inputLower.includes(k.toLowerCase()));
                }
            }

            const keywordMatch = primaryMatch && secondaryMatch;

            if (keywordMatch || isStickyActive) {
                const entryData = {
                    content: entry.content,
                    order,
                    key: entryKey,
                    keys,
                    comment: entry.comment || entry.name || keys[0] || '未命名',
                    isConstant: false,
                    position,
                    sticky,
                    triggeredByKeyword: keywordMatch,
                    triggeredBySticky: isStickyActive && !keywordMatch
                };

                if (keywordMatch) {
                    matched.push(entryData);
                } else {
                    stickyMatched.push(entryData);
                }
            }
        }

        const all = [...constants, ...matched, ...stickyMatched];
        all.sort((a, b) => b.order - a.order);

        return all.slice(0, maxEntries);
    }

    /**
     * 获取所有世界书列表
     */
    listWorldBooks() {
        if (!fs.existsSync(this.worldsDir)) {
            return [];
        }

        const files = fs.readdirSync(this.worldsDir);
        return files.filter(f => f.endsWith('.json'));
    }

    /**
     * 清除缓存
     */
    clearCache() {
        this.cache.clear();
    }

    /**
     * 获取当前加载的世界书
     */
    getCurrentWorldBook() {
        if (!this.currentWorldBook) {
            return null;
        }
        return {
            name: this.currentWorldBookName,
            entries: this.currentWorldBook.entries ? Object.keys(this.currentWorldBook.entries).length : 0
        };
    }

    /**
     * 加载世界书
     */
    loadWorldBook(filename) {
        const filePath = path.join(this.worldsDir, filename);
        if (!fs.existsSync(filePath)) {
            throw new Error(`世界书文件不存在: ${filePath}`);
        }

        const worldBook = safeJsonParse(fs.readFileSync(filePath, 'utf8'), 10 * 1024 * 1024);
        this.currentWorldBook = worldBook;
        this.currentWorldBookName = filename.replace('.json', '');
        this.cache.set(filename, worldBook);
        
        return {
            name: this.currentWorldBookName,
            entries: worldBook.entries ? Object.keys(worldBook.entries).length : 0
        };
    }

    /**
     * 扫描世界书目录（刷新列表）
     */
    scanWorldBooks() {
        this.clearCache();
        return this.listWorldBooks();
    }

    /**
     * 查找匹配的条目（用于测试）
     */
    findMatchingEntries(text) {
        if (!this.currentWorldBook) {
            return [];
        }
        return this.matchEntries(this.currentWorldBook, text);
    }
}

File: internal/chat/worldbook.go
package chat

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// 本文件移植 Node src/worldbook.js：世界书读取与条目匹配
// （keys/secondary_keys + selectiveLogic、constant、sticky、order、position）。

// worldBookEntry 是世界书条目的宽容解析形状。
type worldBookEntry struct {
	UID            any      `json:"uid"`
	ID             any      `json:"id"`
	Keys           []string `json:"-"`
	SecondaryKeys  []string `json:"-"`
	Content        string
	Comment        string
	Order          float64
	Position       int
	Sticky         float64
	Constant       bool
	Enabled        bool
	Disable        bool
	SelectiveLogic float64
	EntryKey       string
}

type worldBook struct {
	Entries []map[string]any `json:"entries"`
}

// readWorldBook 对齐 Node WorldBookManager.readWorldBook 的文件名匹配链。
func readWorldBook(dataDir string, characterName string) (*worldBook, string, error) {
	if characterName == "" {
		return nil, "", nil
	}
	worldsDir := filepath.Join(dataDir, "worlds")
	candidates := []string{
		characterName + "’s Lorebook.json",
		characterName + "'s Lorebook.json",
		characterName + " Lorebook.json",
		characterName + ".json",
	}
	for _, name := range candidates {
		path := filepath.Join(worldsDir, name)
		if book, err := loadWorldBookFile(path); err == nil {
			return book, name, nil
		}
	}
	entries, err := os.ReadDir(worldsDir)
	if err != nil {
		return nil, "", nil
	}
	for _, file := range entries {
		if file.IsDir() || !strings.HasSuffix(file.Name(), ".json") {
			continue
		}
		if !strings.Contains(file.Name(), characterName) {
			continue
		}
		path := filepath.Join(worldsDir, file.Name())
		if book, err := loadWorldBookFile(path); err == nil {
			return book, file.Name(), nil
		}
	}
	return nil, "", nil
}

func loadWorldBookFile(path string) (*worldBook, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var book worldBook
	if err := json.Unmarshal(data, &book); err != nil {
		return nil, err
	}
	return &book, nil
}

func worldBookKeys(value any) []string {
	switch typed := value.(type) {
	case []any:
		keys := []string{}
		for _, item := range typed {
			if text, ok := item.(string); ok && strings.TrimSpace(text) != "" {
				keys = append(keys, text)
			}
		}
		return keys
	case string:
		if strings.TrimSpace(typed) != "" {
			return []string{typed}
		}
	}
	return nil
}

func worldBookEnabled(value any, fallback bool) bool {
	switch typed := value.(type) {
	case nil:
		return fallback
	case bool:
		return typed
	case string:
		normalized := strings.ToLower(strings.TrimSpace(typed))
		switch normalized {
		case "":
			return fallback
		case "false", "0", "off", "no":
			return false
		case "true", "1", "on", "yes":
			return true
		}
		return typed != ""
	case float64:
		return typed != 0
	default:
		return fallback
	}
}

func worldBookNumber(value any, fallback float64) float64 {
	switch typed := value.(type) {
	case float64:
		return typed
	case string:
		parsed, err := strconv.ParseFloat(strings.TrimSpace(typed), 64)
		if err == nil {
			return parsed
		}
	case int:
		return float64(typed)
	case bool:
		if typed {
			return 1
		}
		return 0
	}
	return fallback
}

// resolveWorldBookPosition 对齐 Node normalizePosition（数字仅 1 为 after，其余为 0）。
func resolveWorldBookPosition(value any) int {
	if text, ok := value.(string); ok {
		normalized := strings.ToLower(strings.TrimSpace(text))
		switch normalized {
		case "1", "after", "after_char", "after_character", "after_description", "post", "post_history", "post-history":
			return 1
		case "0", "before", "before_char", "before_character", "before_description", "system", "pre_history", "pre-history":
			return 0
		}
	}
	// Node 数字分支：Number.isFinite ? (===1?1:0)；非 1（含 2/-1）一律落 0
	if worldBookNumber(value, 0) == 1 {
		return 1
	}
	return 0
}

// matchedWorldBookEntry 是匹配后的条目（对齐 Node matchEntries 输出）。
type matchedWorldBookEntry struct {
	Content            string
	Order              float64
	Key                string
	Keys               []string
	Comment            string
	Position           int
	Sticky             float64
	IsConstant         bool
	TriggeredByKeyword bool
	TriggeredBySticky  bool
}

// matchWorldBookEntries 对齐 Node matchEntries：常驻 + 关键词 + 粘性，按 order 降序截断。
func matchWorldBookEntries(book *worldBook, inputText string, maxEntries int, stickyKeys map[string]bool) []matchedWorldBookEntry {
	if book == nil || len(book.Entries) == 0 {
		return nil
	}
	if maxEntries <= 0 {
		maxEntries = 10
	}
	if stickyKeys == nil {
		stickyKeys = map[string]bool{}
	}
	inputLower := strings.ToLower(inputText)
	constants := []matchedWorldBookEntry{}
	matched := []matchedWorldBookEntry{}
	stickyMatched := []matchedWorldBookEntry{}

	for _, raw := range book.Entries {
		if !worldBookEnabled(raw["enabled"], true) || worldBookEnabled(raw["disable"], false) {
			continue
		}
		keys := worldBookKeys(orValue(raw["keys"], raw["key"]))
		entryKey := entryKeyOf(raw, keys)
		sticky := worldBookNumber(raw["sticky"], 0)
		order := worldBookNumber(orValue(raw["order"], raw["insertion_order"]), 0)
		position := resolveWorldBookPosition(raw["position"])

		if worldBookEnabled(raw["constant"], false) {
			constants = append(constants, matchedWorldBookEntry{
				Content: stringOf(raw["content"]), Order: order, Key: entryKey,
				IsConstant: true, Position: position,
			})
			continue
		}

		isStickyActive := stickyKeys[entryKey]
		secondaryKeys := worldBookKeys(orValue(raw["secondary_keys"], raw["keysecondary"]))
		primaryMatch := false
		secondaryMatch := len(secondaryKeys) == 0
		for _, key := range keys {
			if key != "" && strings.Contains(inputLower, strings.ToLower(key)) {
				primaryMatch = true
				break
			}
		}
		if primaryMatch && len(secondaryKeys) > 0 {
			logic := worldBookNumber(raw["selectiveLogic"], 0)
			switch int(logic) {
			case 0:
				for _, k := range secondaryKeys {
					if k != "" && strings.Contains(inputLower, strings.ToLower(k)) {
						secondaryMatch = true
						break
					}
				}
			case 1:
				allContain := true
				for _, k := range secondaryKeys {
					if k == "" || !strings.Contains(inputLower, strings.ToLower(k)) {
						allContain = false
						break
					}
				}
				secondaryMatch = !allContain
			case 2:
				secondaryMatch = true
				for _, k := range secondaryKeys {
					if k != "" && strings.Contains(inputLower, strings.ToLower(k)) {
						secondaryMatch = false
						break
					}
				}
			case 3:
				secondaryMatch = true
				for _, k := range secondaryKeys {
					if k == "" || !strings.Contains(inputLower, strings.ToLower(k)) {
						secondaryMatch = false
						break
					}
				}
			}
		}
		keywordMatch := primaryMatch && secondaryMatch
		if !keywordMatch && !isStickyActive {
			continue
		}
		entry := matchedWorldBookEntry{
			Content: stringOf(raw["content"]), Order: order, Key: entryKey, Keys: keys,
			Comment:  firstNonEmptyString(stringOf(raw["comment"]), stringOf(raw["name"]), firstOf(keys), "未命名"),
			Position: position, Sticky: sticky,
			TriggeredByKeyword: keywordMatch, TriggeredBySticky: isStickyActive && !keywordMatch,
		}
		if keywordMatch {
			matched = append(matched, entry)
		} else {
			stickyMatched = append(stickyMatched, entry)
		}
	}

	all := append(append(constants, matched...), stickyMatched...)
	sort.SliceStable(all, func(i, j int) bool { return all[i].Order > all[j].Order })
	if len(all) > maxEntries {
		all = all[:maxEntries]
	}
	return all
}

func entryKeyOf(raw map[string]any, keys []string) string {
	for _, field := range []string{"uid", "id"} {
		if value, ok := raw[field]; ok && value != nil {
			if text := stringOf(value); text != "" {
				return text
			}
			if number, ok := value.(float64); ok {
				return fmt.Sprintf("%v", number)
			}
		}
	}
	if len(keys) > 0 && keys[0] != "" {
		return keys[0]
	}
	return firstNonEmptyString(stringOf(raw["comment"]), stringOf(raw["name"]), "unknown")
}

func orValue(primary any, secondary any) any {
	if primary != nil {
		return primary
	}
	return secondary
}

func stringOf(value any) string {
	if text, ok := value.(string); ok {
		return text
	}
	return ""
}

func firstOf(values []string) string {
	if len(values) > 0 {
		return values[0]
	}
	return ""
}

func firstNonEmptyString(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}
	return ""
}
